File: trunk/Lite/src/main/org/xidea/lite/parse/ParseContext.java
package org.xidea.lite.parse;

import java.io.IOException;
import java.net.URI;
import java.util.Collection;
import java.util.List;
import java.util.Map;


import org.w3c.dom.Document;
import org.xidea.lite.impl.ParseContextImpl;
import org.xml.sax.SAXException;

/**
 * @see ParseContextImpl
 */
public interface ParseContext extends ResultContext, ParserHolder {
	public String NS_CORE = "http://www.xidea.org/lite/core";
	// 默认值：link|input|meta|img|br|hr
//	public String FEATURE_HTML_LEAF = "http://www.xidea.org/lite/features/html-leaf";
	// 默认值为空
	public String FEATURE_HTML_TRIM = "http://www.xidea.org/lite/features/html-trim";
	// 默认值为空
	public String FEATURE_HTML_JAVASCRIPT_COMPRESSOR = "http://www.xidea.org/lite/features/html-javascript-compressor";
	// 默认值为utf-8
	public String FEATURE_ENCODING = "http://www.xidea.org/lite/features/output-encoding";
	// 默认值为 text/html
	public String FEATURE_MIME_TYPE = "http://www.xidea.org/lite/features/output-mime-type";
	// 默认值为 null,可以采用相对路径模式：layout='./layout.xhtml'
	public String FEATURE_LAYOUT = "http://www.xidea.org/lite/features/config-layout";

	/**
	 * 给出文件内容或url，解析模版源文件
	 */
	public void parse(Object source);

	/**
	 * 获取当前正在解析的模版URI
	 * @return
	 */
	public URI getCurrentURI();

	/**
	 * 获取当前正在解析的模版URI
	 * 同事将该url记录在资源列表中
	 * @return
	 */
	public void setCurrentURI(URI currentURI);
	/**
	 * 添加（记录）解析相关的资源
	 * @param resource
	 */
	public void addResource(URI resource);
	
	public Collection<URI> getResources();
	
	/**
	 * 记录一下编译上下文状态
	 * @param key
	 * @param value
	 */
	public void setAttribute(Object key, Object value);
	public <T> T getAttribute(Object key);
	/**
	 * 当前代码类型；
	 * 直接使用Template中的常量定义：
	 * Template.XT_TYPE
	 * Template.XA_TYPE
	 * Template.EL_TYPE
	 * @return
	 */
	public int getTextType();
	/**
	 * 是否保留空白（默认为false）
	 * @return
	 */
	public boolean isReserveSpace();
	public void setReserveSpace(boolean keepSpace);
	/**
	 * 给出文件内容或url，解析模版源文件
	 */
	public List<Object> parseText(String text, int textType);

	/**
	 * 如果file相于根目录（/path/...），以base作为根目录处理 否则以parentURI，或者base作为parent直接new
	 * URL处理。
	 * 
	 * @param file
	 * @param parentURI
	 * @see org.xidea.lite.impl.ParseContextImpl#createURI
	 * @return
	 */
	public URI createURI(String file);

	/**
	 * 装载指定XML文档。
	 * loadXML 一般不会设计为采用{@link ParseContext#loadText(URI)} 的结果，
	 * 因为xml的编码有自己的规范。
	 * @param uri
	 * @return
	 * @throws SAXException
	 * @throws IOException
	 */
	public Document loadXML(URI uri) throws SAXException, IOException;
	/**
	 * 装载指定资源文本。
	 * 
	 * @param uri
	 * @return
	 * @throws IOException
	 */
	public String loadText(URI uri) throws IOException;
	/**
	 * 记录一下编译上下文特征变量，该对象不可被修改
	 * 
	 */
	public String getFeature(String key);

	/**
	 * 只读，外部不允许修改
	 * @return featureMap {url,value}
	 */
	public Map<String, String> getFeatureMap();
}

File: trunk/Lite/web/scripts/org/xidea/lite/parse/parse-context.js
if(typeof require == 'function'){
var ResultContext=require('./result-context').ResultContext;
var ParseConfig=require('./config').ParseConfig;
var URI=require('org/xidea/lite/util/resource').URI;
var loadLiteXML=require('org/xidea/lite/util/xml').loadLiteXML;
}/*
 * List Template
 * License LGPL(您可以在任何地方免费使用,但请不要吝啬您对框架本身的改进)
 * http://www.xidea.org/project/lite/
 * @author jindw
 * @version $Id: template.js,v 1.4 2008/02/28 14:39:06 jindw Exp $
 */
var defaultBase = new URI("lite:///");
//add as default
/**
 * 模板解析上下文对象实现
 */
function ParseContext(config,path){
	config = config || new ParseConfig();
    this.config = config;
	this.currentURI = defaultBase;
	this.featureMap = config.getFeatureMap(path);
    this.textType=0;
	this._path = path;
	this._attributeMap = [[],[],{}]
    this._result = new ResultContext();
	this._context = this;
	this._result._context = this;
	this._resources = [];
	initializeParser(this,config.getExtensionMap(path));
}
/**
 * 初始化上下文
 * @arguments 链顶插入的解析器列表（第一个元素为初始化后的链顶解析器，以后类推）
 */
function initializeParser(context,extensionMap){
	var extensionParser = new ExtensionParser();
	for(var ns in extensionMap){
		var exts = extensionMap[ns];
		for(var len = exts.length,i=0;i<len;i++){
			extensionParser.addExtension(ns,exts[i])
		}
	}
	context._nodeParsers = [parseTextLeaf,parseDefaultXMLNode,parseExtension];
	context._textParsers = [extensionParser];
	context._extensionParser = extensionParser;
    context._topChain = buildTopChain(context);
}
function parseExtension(node,context,chain){//extension
	return context._extensionParser.parse(node,context,chain);
}
function parseTextLeaf(text,context){
	if(typeof text == 'string'){
		return parseText(text,context,context._textParsers)
	}else{
		console.error("未知节点类型",typeof text,text)
		//chain.next(text);
	}
}
ParseContext.prototype = {
	parseText:function(source, textType) {
		switch(textType){
		case XA_TYPE :
	    case XT_TYPE :
	    case EL_TYPE :
	        break;
	    default:
			console.error("未知编码模式："+textType)
			throw new Error();
		}
		
		var mark = this.mark();
		var oldType = this.textType;
		this._context.textType = textType;
		parseTextLeaf(source,this);
		this._context.textType = oldType;
		var result = this.reset(mark);
		return result;
	},
    /**
     * 调用解析链顶解析器解析源码对象
     * @param 文本源代码内容或xml源代码文档对象。
     * @public
     * @abstract
     */
	parse:function(source) {
		var type = source.nodeType;
		if(type>0){//xml
			//console.info(len,source && source.xml)
			this._topChain.next(source);
		}else{//text
			if(source instanceof URI){
				source = this.loadXML(source);
				if(typeof source == 'string'){
					source=source.replace(/#.*[\r\n]*/,'');
				}
			}
			if(typeof source != 'string'){
				//NodeList
				var len = source.length;
				var nodeType = source.nodeType;
				
				if(nodeType === undefined && typeof source.item != 'undefined'){//NodeList
					if(len === 0){
						return;
					}
					for(var i = 0;i<len;i++){
						this._topChain.next(source.item(i));
					}
					return;
				}
			}
			this._topChain.next(source);
		}
		
		
	},
    createURI:function(path) {
    	//console.error(path,this.currentURI,this.config.root)
    	var base = this.config.root.toString();
    	if(path.indexOf(base) ==0){
    		path = path.substring(base.length-1);
    	}
    	var cu = this.currentURI;
    	if(cu){
    		//if(cu.scheme == 'data'){
    		//	return new URI(cu);
    		//}else{
    		//console.log(path,cu)
    		//console.log('???'+cu.resolve(path))
    		return cu.resolve(path);
    		//}
    	}else{
    		path= path.replace(/^[\\\/]/,'./');// /xxx=>./xxx
    		//console.warn(defaultBase+'',path,defaultBase.resolve(path)+'',defaultBase.authority)
    		
    		//console.log(path,defaultBase)
    		//console.log('###'+defaultBase.resolve(path))
    		return defaultBase.resolve(path);
    	}
    	
    	
    },
    loadText:function(uri){
    	//only for java
    	if(uri.scheme == 'lite'){
    		var path = uri.path+(uri.query||'');
    		path = path.replace(/^\//,'./')
    		uri = this.config.root.resolve(path);
    	}
    	var xhr = new XMLHttpRequest();
	    xhr.open("GET",uri,false)
	    xhr.send('');
	    ////text/xml,application/xml...
	    return xhr.responseText;
    },
    loadXML:function(path){
    	var t1 = +new Date();
    	if(path instanceof URI){
    	}else{
    		if(/^\s*</.test(path)){
    			return loadLiteXML(path,this.config.root)
    		}else{
    			path = new URI(path)
    		}
    	}
    	this.setCurrentURI(path);
    	var doc = loadLiteXML(path,this.config.root);
    	this._context._loadTime+=(new Date()-t1);
    	return doc;
    },
    openStream:function(uri){
//    	//only for java
//    	if(uri.scheme == 'lite'){
//    		var path = uri.path+(uri.query||'');
//    		path = path.replace(/^\//,'./')
//    		uri = this.config.root.resolve(path);
//    	}
//    	return Packages.org.xidea.lite.impl.ParseUtil.openStream(uri)
		throw new Error("only for java");
    },
	setAttribute:function(key,value){
		setByKey(this._context._attributeMap,key,value)
	},
	getAttribute:function(key){
		return getByKey(this._context._attributeMap,key)
	},
	addNodeParser:function(np){
		this._nodeParsers.push(np);
	},
	addTextParser:function(tp){
		this._textParsers.push(tp);
	},
	addExtension:function(ns,pkg){
		this._extensionParser.addExtension(ns,pkg);
	},
	getFeature:function(key){
		return this.featureMap[key];
	},
	getFeatureMap:function(){
		return this.featureMap;
	},
    setCurrentURI:function(uri){
    	this._context.addResource(uri=new URI(uri));
    	this._context.currentURI = uri;
    },
    addResource:function(uri){
    	for(var rs = this._resources, i=0;i<rs.length;i++){
    		if(rs[i]+'' == uri){
    			return ;
    		}
    	}
    	this._resources.push(uri);
    },
    getResources:function(){
    	return this._resources;
    },
    createNew:function(){
    	var nc = new ParseContext(this.config,this.currentURI);
    	nc.featureMap = this.featureMap;
    	nc._resources = this._resources;
    	return nc;
    },
    _loadTime :0
}
var rm = ResultContext.prototype;
for(var n in rm){
	if(rm[n] instanceof Function){
		ParseContext.prototype[n] = buildResultWrapper(n);
	}
}
function buildResultWrapper(n){
	return function(){
		var result = this._result;
		return result[n].apply(result,arguments)
	}
}
if(typeof require == 'function'){
exports.ParseContext=ParseContext;
var buildTopChain=require('./parse-chain').buildTopChain;
var ExtensionParser=require('./extension-parser').ExtensionParser;
var Extension=require('./extension').Extension;
var parseDefaultXMLNode=require('./xml-default-parser').parseDefaultXMLNode;
var parseText=require('./text-parser').parseText;
var getByKey=require('org/xidea/lite/util/kv').getByKey;
var setByKey=require('org/xidea/lite/util/kv').setByKey;
var XA_TYPE=require('org/xidea/lite/impl/template-token').XA_TYPE;
var EL_TYPE=require('org/xidea/lite/impl/template-token').EL_TYPE;
var XT_TYPE=require('org/xidea/lite/impl/template-token').XT_TYPE;
}